### ISSUE TITLE
new language model for pos tagging

### ISSUE THREAD

**Issue by junaidamk** [zalandoresearch/flair#643]
I tried to implement pos tagger for Malayalam(Indian Language)using flair. i have 31000 tagged instances. is it enough to train a tagger with pretrained embedding? my code snippet is like this

from flair.data import TaggedCorpus
from flair.data_fetcher import NLPTaskDataFetcher, NLPTask
from flair.embeddings import TokenEmbeddings, WordEmbeddings, StackedEmbeddings
from typing import List
# 1. get the corpus
 #define columns
columns = {0: 'text', 1: 'pos'}

# this is the folder in which train, test and dev files reside
data_folder = '.'

# retrieve corpus using column format, data folder and the names of the train, dev and test files
corpus: TaggedCorpus = NLPTaskDataFetcher.load_column_corpus(data_folder, columns,
                                                              train_file='train.txt',
                                                              test_file='test.txt',
                                                              dev_file='dev.txt')

print(corpus)
tag_type = 'pos'
# 3. make the tag dictionary from the corpus
tag_dictionary = corpus.make_tag_dictionary(tag_type=tag_type)
print(tag_dictionary.idx2item)

output is below
Better speed can be achieved with apex installed from https://www.github.com/nvidia/apex.
2019-04-04 10:37:23,717 Reading data from .
2019-04-04 10:37:23,718 Train: train.txt
2019-04-04 10:37:23,719 Dev: dev.txt
2019-04-04 10:37:23,720 Test: test.txt
TaggedCorpus: 24000 train + 3495 dev + 3342 test sentences
[b'<unk>', b'O', b'N_NN', b'V_VM_VNF', b'N_NNP', b'PSP', b'V_VAUX', b'RD_PUNC', b'CC_CCD', b'JJ', b'V_VM_VF', b'QT_QTC', b'N_NST', b'CC_CCS', b'QT_QTF', b'PR_PRF', b'DM_DMD', b'PR_PRP', b'RP_NEG', b'V_VM_VINF', b'RB', b'PR_PRL', b'CC_CCS_UT', b'RD_SYM', b'RP_INTF', b'RD_RDF', b'DM_DMQ', b'RP_RPD', b'QT_QTO', b'DM_DMR', b'RP_CL', b'PR_PRC', b'V_VM', b'RP_INJ', b'V_VN', b'cfp[r4ewi\xe0\xb4\xa8\xe0\xb5\x82\xe0\xb4\xb1\xe0\xb5\x8d\xe0\xb4\xb1\xe0\xb4\xbf\xe0\xb4\x92\xe0\xb4\xa8\xe0\xb5\x8d\xe0\xb4\xa8\xe0\xb4\xbf\xe0\xb4\xb2\xe0\xb5\x8d\xe2\x80\x8d', b'PR_PRQ', b'RD_ECH', b'<START>', b'<STOP>']
 Here i got extra tags other than i used such as b'cfp[r4ewi\xe0\xb4\xa8\xe0\xb5\x82\xe0\xb4\xb1\xe0\xb5\x8d\xe0\xb4\xb1\xe0\xb4\xbf\xe0\xb4\x92\xe0\xb4\xa8\xe0\xb5\x8d\xe0\xb4\xa8\xe0\xb4\xbf\xe0\xb4\xb2\xe0\xb5\x8d\xe2\x80\x8d,
i couldnt see anything like this in our corpus? actually what cause of this?

**To Reproduce**
Steps to reproduce the behavior (e.g. which model did you train? what parameters did you use? etc.).

**Expected behavior**

**Screenshots**

**Environment (please complete the following information):**
 - OS [e.g. iOS, Linux]:
 - Version [e.g. flair-0.3.2]:

**Additional context**

**Comment by stefan-it** [comment]
@junaidamk Could you provide a sample from your dataset?  These `b'cfp[r4ewi\xe0\xb4\xa8\xe0\xb5\x82\xe0\xb4\xb1\xe....` tags look really weird. I guess there's a "wrong" line in your dataset format 🤔

**Comment by stefan-it** [comment]
Are you using this dataset: https://github.com/rbvijay/POS-RNN-malayalam/tree/master/resources (`mal_pos.*`)?

**Comment by junaidamk** [comment]
No sir, I have used a corpus under INDIAN LANGUAGE CORPORA INITIATVE(ILCI- PHAAE -2) of meity, India. The format is same as above link..I have checked my corpus couldn't find anything like this..

**Comment by stefan-it** [comment]
I tried to find the corpus; is it that one:

https://tdil-dc.in/index.php?option=com_download&task=showresourceDetails&toolid=1886&lang=en

?

**Comment by junaidamk** [comment]
Yes..I converted this in 2 column format

**Comment by junaidamk** [comment]
this data set is enough for my experiments? Any suggestions please ....
Also I tried to convert fastext malayalam crawl embedding to gensim format. But I got only .npy file? 
What are the fine tune parameters for my experiment with this dataset.

**Comment by stefan-it** [comment]
Hi @junaidamk, yes 24K sentences are totally ok to train on. See e.g. the PoS tagging datasets for various languages on the Universal Dependencies website: https://universaldependencies.org/#current-ud-languages and you'll find some languages with even lower training examples.

To your other questions:

Here's a code snippet that I use in my experiments, when I try to load word embeddings from the FastText page:

a) Download them:

```bash
wget https://dl.fbaipublicfiles.com/fasttext/vectors-crawl/cc.ml.300.vec.gz
gunzip cc.ml.300.vec.gz
```

b) Load them in Python

```python
word_vectors = gensim.models.KeyedVectors.load_word2vec_format('cc.ml.300.vec', binary=False)
word_vectors.save('cc.ml.300.vec.gensim')

custom_embedding = WordEmbeddings('cc.ml.300.vec.gensim')
```

Hyperparameter fine-tuning can be done on several places:

```
tagger: SequenceTagger = SequenceTagger(hidden_size=256,
                                        embeddings=embeddings,
                                        tag_dictionary=tag_dictionary,
                                        tag_type=tag_type,
use_crf=True)
```

Here you could fine-tune the number of hidden sizes of the Bi-LSTM.

The `.train()` function of the `ModelTrainer` allows you to also modify hyperparameters:

```python
trainer.train('resources/taggers/example-ner', EvaluationMetric.MICRO_F1_SCORE, learning_rate=0.1, mini_batch_size=32,
max_epochs=20, test_mode=True)
```

Paramaters like learning rate, batch size or maximum number of epochs can be adjusted. A full (hyper-) parameter list can be found [here](https://github.com/zalandoresearch/flair/blob/master/flair/trainers/trainer.py#L40).

You could even run a hyperparameter search. For that I just point you to the [model tuning](https://github.com/zalandoresearch/flair/blob/master/resources/docs/TUTORIAL_8_MODEL_OPTIMIZATION.md#tutorial-8-model-tuning) section.

I would really like to help you with the dataset parsing. But unfortunately, the registration process for the "Malayalam Monolingual Text Corpus ILCI-II" corpus is so complicated, so I couldn't do any conversion. For debugging I would write a small Python script that a) parses your training, development and test set line-by-line and column-wise and b) output suspicious lines that include these "b'cfp[r4ew" occurences (on the label column). Then you could remove these lines from your dataset.

**Comment by junaidamk** [comment]
I have already done as you mentioned..
word_vectors = gensim.models.KeyedVectors.load_word2vec_format('cc.ml.300.vec', binary=False)
word_vectors.save('cc.ml.300.vec.gensim')

But this produces 2 file .npy and vec. When I opened vec file is empty...why?

**Comment by junaidamk** [comment]
> Hi @junaidamk, yes 24K sentences are totally ok to train on. See e.g. the PoS tagging datasets for various languages on the Universal Dependencies website: https://universaldependencies.org/#current-ud-languages and you'll find some languages with even lower training examples.
> 
> To your other questions:
> 
> Here's a code snippet that I use in my experiments, when I try to load word embeddings from the FastText page:
> 
> a) Download them:
> 
> ```shell
> wget https://dl.fbaipublicfiles.com/fasttext/vectors-crawl/cc.ml.300.vec.gz
> gunzip cc.ml.300.vec.gz
> ```
> 
> b) Load them in Python
> 
> ```python
> word_vectors = gensim.models.KeyedVectors.load_word2vec_format('cc.ml.300.vec', binary=False)
> word_vectors.save('cc.ml.300.vec.gensim')
> 
> custom_embedding = WordEmbeddings('cc.ml.300.vec.gensim')
> ```
> 
> Hyperparameter fine-tuning can be done on several places:
> 
> ```
> tagger: SequenceTagger = SequenceTagger(hidden_size=256,
>                                         embeddings=embeddings,
>                                         tag_dictionary=tag_dictionary,
>                                         tag_type=tag_type,
> use_crf=True)
> ```
> 
> Here you could fine-tune the number of hidden sizes of the Bi-LSTM.
> 
> The `.train()` function of the `ModelTrainer` allows you to also modify hyperparameters:
> 
> ```python
> trainer.train('resources/taggers/example-ner', EvaluationMetric.MICRO_F1_SCORE, learning_rate=0.1, mini_batch_size=32,
> max_epochs=20, test_mode=True)
> ```
> 
> Parameters like learning rate, batch size or maximum number of epochs can be adjusted. A full (hyper-) parameter list can be found [here](https://github.com/zalandoresearch/flair/blob/master/flair/trainers/trainer.py#L40).
> 
> You could even run a hyperparameter search. For that I just point you to the [model tuning](https://github.com/zalandoresearch/flair/blob/master/resources/docs/TUTORIAL_8_MODEL_OPTIMIZATION.md#tutorial-8-model-tuning) section.
> 
> I would really like to help you with the dataset parsing. But unfortunately, the registration process for the "Malayalam Monolingual Text Corpus ILCI-II" corpus is so complicated, so I couldn't do any conversion. For debugging I would write a small Python script that a) parses your training, development and test set line-by-line and column-wise and b) output suspicious lines that include these "b'cfp[r4ew" occurences (in the label column). Then you could remove these lines from your dataset.

I parsed the data and removed these lines from dataset.........Thank you for your support........

**Comment by junaidamk** [comment]
Now I got solution for fasttext to gensim conversion also....thank you...

**Comment by alanakbik** [comment]
Hello @junaidamk hope everything runs now - do let us know how well it works :) I'll close the issue for now, but feel free to reopen if you have more questions!

**Comment by junaidamk** [comment]
yes..the logfile with fast
text embedding is attached here..suggestions please.....
[training.log](https://github.com/zalandoresearch/flair/files/3115478/training.log)
